test/ServerSpec.js

```diff
@@ -65,4 +67,115 @@ describe('Server Spec', () => {
         .end(done);
     });
   });
+
+  describe('/search/:market/:checkin/:checkout', () => {
+    it('Should retrieve all San Francisco listings available for checkin 2017-11-12 checkout 2017-11-13', (done) => {
+      request(server)
+        .get('/search/San%20Francisco/2017-11-12/2017-11-13')
+        .expect('Content-Type', /json/)
+        .expect(200)
+        .expect((response) => {
+          const availableListings = response.body.filter(result => result.inventory_date.split('T')[0] === '2017-11-12');
+          expect(availableListings.length).to.equal(response.body.length);
+        })
+        .end(done);
+    });
+
+    it('Should retrieve no San Francisco listings available for checkin 2017-10-02 checkout 2017-10-03', (done) => {
+      request(server)
+        .get('/search/San%20Francisco/2017-10-02/2017-10-03')
+        .expect('Content-Type', /json/)
+        .expect(200)
+        .expect((response) => {
+          expect(response.body.length).to.equal(0);
+        })
+        .end(done);
+    });
+
+    it('Should retrieve all San Francisco listings available for checkin 2017-10-19 and checkout 2017-10-24', (done) => {
+      const availableListings = new Set();
+      const stayDates = {
+        '2017-10-19': 0, '2017-10-20': 0, '2017-10-21': 0, '2017-10-22': 0, '2017-10-23': 0,
+      };
+      request(server)
+        .get('/search/San%20Francisco/2017-10-19/2017-10-24')
+        .expect('Content-Type', /json/)
+        .expect(200)
+        .expect((response) => {
+          response.body.forEach((result) => {
+            availableListings.add(result.listing_id);
+            stayDates[result.inventory_date.split('T')[0]] += 1;
+          });
```

Comment:
I don't quite understand the flow of this logic. So for every entry you get back, you are adding the listing (uniquely) into available listings (so that listings for all 3 days are each added only once). Then you increment the number of stays for each listings on that date.

In the last one, you are saying that you expect the number of available stay days for that date to equal the number of listings you have. I don't quite follow this logic?

Reply:
Consider the following example: Search for listings available for check-in/check-out October 28/October 30. This is a two night stay that requires availability on October 28 and October 29. In my database, I have the following availability:

listing_id-inventory_date
1-October 28
1-October 29
2-October 28
3-October 28
4-October 28
4-October 29

Listings 1, 2, 3, 4 match the criteria of having availability records that fall between on October 28 and October 29, inclusive. But only listings 1 and 4 are actually available for the entire length of stay. If all 4 listings were returned in the search results, I would have the following:
`availableListings.size = 4`
`stayDates['2017-10-28'] = 4`
`stayDates['2017-10-29'] = 2`

This means that I got 4 listings back but not all of them are available for all days during the stay. So this should fail the test.